Contributing-errors.md

```diff
@@ -0,0 +1,163 @@
+  - Alternative situations exist where the developer is using trial an error
+    and possibly looking at samples to get it working for their usecase.
+    In this situation they will likely be more accepting of errors,
+    but hopeful the errors will ease their burden and reveal how to resolve the issue.
+  - In both these cases
+    developers have not built the kind of rapport with the platform where they are likely to contribute fixes or dig deep
+    unless they are instructed to by their employer.
+    Failed exploration might lead to investigation into alternative products
+- The developer is implementing our product
+- The developer is productionizing our product
+- The developer is contributing to our product
+
+### Principles
+
+In all these cases some truth remain, errors should be **available**, **useful**, and **actionable**.
+
+#### *available* errors need to be available to developers
+
+The worst situation for external developers is when errors are unavailable.
+The immediate reaction is that the application is severely broken.
+Consider a situation where an application takes user input
+and attempts to write it to a database.
+If no error is provided and the data is not revealed in a result
+then the user might think the data is lost.
+This kind of thinking can result in the failed adoption of the product.
+
+Making errors available is the simplest solution by means of passing errors up,
+but I would go one step further and make sure that the error passed up is scoped to the product at hand
+and not just a byproduct of the stdlib.
+In this way errors are available and easy to identify when stacktraces are unavailable.
+
+> All returned errors must be exported, either as a type or constant.
+
+#### *useful* - errors need to be readable for the target developer language
+
+OpenTDF is targeting English speaking developers,
+but with the desire that it is useful internationally.
+It isn’t realistic to think that we could support internationalization today with regards to errors,
+but if there are ways we can set ourselves up to enrich our errors or expose them some documentation glossary where internationalization is more easily supported.
+
+A good illustration of this is when English speakers use libraries like Ant Design (Alibaba backed frontend design system).
+The value of Ant Design as a product excuses some of the grammar and language barriers for native English speakers,
+but anecdotally, it seems that English-as-a-second-language (ESL) speakers (who also don’t know Chinese) tend to favor English-driven tooling.
+
+Errors should use proper grammar and spelling.
+Usefulness and accuracy of errors is reenforced by the quality of the grammar and spelling of errors.
+This is especially true when we are striving for international usage while not having the capacity for internationalization.
+
+Errors should be idiomatic with the language.
```

Comment:
Can we specify a noun -> adjective structure in error naming? For example, we would prefer `ErrCredentialsMissing` and `ErrCredentialsInvalid` instead of `ErrMissingCredentials` or `ErrInvalidCredentials`. This will help with grouping similar errors to reduce cognitive load.